python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+        tensor([[2., 2., 2., 2., 2.],
+                [2., 2., 2., 2., 2.],
+                [2., 2., 2., 2., 2.]])
+        """
+        pass
+
+    def group_apply_edges(self, group_by, func, edges=ALL, inplace=False):
+        """Group the edges by nodes and apply the function of the grouped
+        edges to update their features.  The edges are of the same edge type
+        (hence having the same source and destination node type).
+
+        Notes
+        -----
+        An error is raised if the graph contains multiple edge types.  Use
+
+            g[src_type, dst_type, edge_type]
+
+        to select edges with type ``(src_type, dst_type, edge_type)``.
+
+        Parameters
+        ----------
+        group_by : str
+            Specify how to group edges. Expected to be either 'src' or 'dst'
+        func : callable
+            Apply function on the edge. The function should be
+            an :mod:`Edge UDF <dgl.udf>`. The input of `Edge UDF` should
+            be (bucket_size, degrees, *feature_shape), and
+            return the dict with values of the same shapes.
+        edges : valid edges type, optional
+            Edges on which to group and apply ``func``. See :func:`send` for valid
+            edges type. Default is all the edges.
+        inplace: bool, optional
+            If True, update will be done in place, but autograd will break.
+        """
+        pass
+
+    # TODO: REVIEW
+    def send(self, edges=ALL, message_func=None):
+        """Send messages along the given edges with the same edge type.
+
+        ``edges`` can be any of the following types:
+
+        * ``int`` : Specify one edge using its edge id.
+        * ``pair of int`` : Specify one edge using its endpoints.
+        * ``int iterable`` / ``tensor`` : Specify multiple edges using their edge ids.
+        * ``pair of int iterable`` / ``pair of tensors`` :
+          Specify multiple edges using their endpoints.
+
+        The UDF returns messages on the edges and can be later fetched in
+        the destination node's ``mailbox``. Receiving will consume the messages.
+        See :func:`recv` for example.
+
+        If multiple ``send`` are triggered on the same edge without ``recv``. Messages
+        generated by the later ``send`` will overwrite previous messages.
+
+        Notes
+        -----
+        An error is raised if the graph contains multiple edge types.  Use
+
+            g[src_type, dst_type, edge_type]
+
+        to select edges with type ``(src_type, dst_type, edge_type)``.
+
+        Consecutive sends on different edge types will also raise an error.
+
+        Parameters
+        ----------
+        edges : valid edges type, optional
+            Edges on which to apply ``message_func``. Default is sending along all
+            the edges.
+        message_func : callable
+            Message function on the edges. The function should be
+            an :mod:`Edge UDF <dgl.udf>`.
```

Comment:
should we support list of message functions?

Reply:
You mean a list of message functions for different edge types?

Reply:
yes

Reply:
Did we agree on having a dictionary of message/reduce functions last time?  I thought we said that the dictionary would look ugly or something.

Reply:
I thought we agreed that we'll allow both. Having a dict of message functions allows us to perform all computation in one API, which has two advantages:
* potentially have better optimizations
* avoid storing many temporary node data in frames.

Reply:
Then what if I do something like this:
```python
g.update_all(
    message_func={
        ('A', 'B', 'AB'): lambda edges: {'x': funcAB(edges)},
        ('C', 'B', 'CB'): lambda edges: {'y': funcCB(edges)}},
    reduce_func={'B': lambda nodes: {'m': funcB(nodes)}})
```
Note that the messages to B nodes do not align.

Typed mailboxes can solve the issue though (i.e. `nodes.typed_mailbox[('A', 'B', 'AB')]['x']` and `nodes.typed_mailbox[('C', 'B', 'CB')]['y']`), so that messages are batched by edge type.  What do you think?

Reply:
should the received messages be stored in `mailbox['x']` and `mailbox['y']`?

Reply:
The problem is that AB and CB could technically return different messages with the same name `x` - even with different shapes.

Reply:
i see. this is problematic. we could simply report an error.

in general, i think we should define large APIs. This principle is used from the very beginning. That's why we have multiple levels of API. We believe large APIs give us more opportunities for optimization in the future.

Reply:
> i see. this is problematic. we could simply report an error.

Why is `typed_mailbox` bad?  Is it simply because we have another API?

Reply:
I'm OK with `typed_mailbox`. I'm a little worried that it may make the programming in reduce very complex. For example, in RGCN, neighbor embeddings of different node types are projected into the same space, so we can do simple reduction in the reduce function. If we use `typed_mailbox`, i'm not sure how to write a simple reduce function for this case. If a model really wants to send messages with different dimensions to the mailbox, it has the option of using a different name for each node type if they want.
We might want to investigate what is the common case. Do most of the models send messages of the same dimension to the destination nodes?